dbms/src/Flash/Coprocessor/StreamingDAGResponseWriter.cpp

```diff
@@ -75,9 +81,29 @@ template <class StreamWriterPtr>
 void StreamingDAGResponseWriter<StreamWriterPtr>::finishWrite()
 {
     if (should_send_exec_summary_at_last)
-        batchWrite<true>();
+    {
+        if (enableFineGrainedShuffle(fine_grained_shuffle_stream_count))
```

Comment:
Is it possible to implement `FineGrainedShuffleWriter` and `BaseShuffleWriter` using c++ templates?

Reply:
We can make this if branch a const expr. But maybe no need to make a new `xxxWriter`? I think there are too many: `StreamingDAGResponseWriter`/`PacketWriter`/`DAGResponseWriter`

Reply:
Make `enable_fine_grained_shuffle` as template argument for `StreamingDAGResponseWriter` and `ExchangeReceiver::readLoop/reactor`.